File: src/main/java/net/insomniakitten/glazed/GlazedClient.java
package net.insomniakitten.glazed;

/*
 *  Copyright 2018 InsomniaKitten
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

import net.insomniakitten.glazed.block.GlassKilnBlock;
import net.minecraft.block.Block;
import net.minecraft.block.BlockPane;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.Minecraft;
import net.minecraft.client.multiplayer.PlayerControllerMP;
import net.minecraft.client.renderer.BufferBuilder;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.block.model.ModelResourceLocation;
import net.minecraft.client.renderer.block.statemap.IStateMapper;
import net.minecraft.client.renderer.block.statemap.StateMapperBase;
import net.minecraft.client.renderer.color.IBlockColor;
import net.minecraft.client.renderer.color.IItemColor;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.client.renderer.texture.TextureMap;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import net.minecraft.client.resources.IResourceManager;
import net.minecraft.client.resources.IResourceManagerReloadListener;
import net.minecraft.entity.Entity;
import net.minecraft.item.Item;
import net.minecraft.util.IStringSerializable;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.World;
import net.minecraftforge.client.event.ColorHandlerEvent;
import net.minecraftforge.client.event.ModelRegistryEvent;
import net.minecraftforge.client.event.RenderWorldLastEvent;
import net.minecraftforge.client.model.ModelLoader;
import net.minecraftforge.fml.client.FMLClientHandler;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import org.lwjgl.opengl.GL11;

import java.util.Arrays;

import static java.util.Objects.requireNonNull;
import static net.minecraft.client.renderer.GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA;
import static net.minecraft.client.renderer.GlStateManager.DestFactor.SRC_COLOR;
import static net.minecraft.client.renderer.GlStateManager.DestFactor.ZERO;
import static net.minecraft.client.renderer.GlStateManager.SourceFactor.DST_COLOR;
import static net.minecraft.client.renderer.GlStateManager.SourceFactor.ONE;
import static net.minecraft.client.renderer.GlStateManager.SourceFactor.SRC_ALPHA;

@SideOnly(Side.CLIENT)
enum GlazedClient implements IResourceManagerReloadListener {
    INSTANCE;

    private static final IStateMapper STATE_MAPPER_BLOCK = new StateMapperBase() {
        @Override
        protected ModelResourceLocation getModelResourceLocation(IBlockState state) {
            final String name = state.getValue(GlazedVariant.PROPERTY).getName() + "_glass_block";
            return new ModelResourceLocation(new ResourceLocation(Glazed.ID, name), "normal");
        }
    };

    private static final IStateMapper STATE_MAPPER_PANE = new StateMapperBase() {
        @Override
        protected ModelResourceLocation getModelResourceLocation(IBlockState state) {
            final String name = state.getValue(GlazedVariant.PROPERTY).getName() + "_glass_pane";
            return new ModelResourceLocation(new ResourceLocation(Glazed.ID, name), String.join(",",
                    "east=" + state.getValue(BlockPane.EAST),
                    "north=" + state.getValue(BlockPane.NORTH),
                    "south=" + state.getValue(BlockPane.SOUTH),
                    "west=" + state.getValue(BlockPane.WEST)
            ));
        }
    };

    private static final IBlockColor BLOCK_COLOR = (state, access, pos, tintIndex) -> {
        if (tintIndex == 0 && access != null && pos != null) {
            return state.getValue(GlazedVariant.PROPERTY).getColor(access, pos);
        }
        return 0xFFFFFFFF;
    };

    private static final IItemColor ITEM_COLOR = (stack, tintIndex) -> {
        final int meta = stack.getMetadata();
        if (tintIndex == 0 && GlazedVariant.isValid(meta)) {
            final Minecraft mc = FMLClientHandler.instance().getClient();
            if (mc.player != null && mc.player.world != null) {
                final World world = mc.player.world;
                final BlockPos pos = mc.player.getPosition();
                return GlazedVariant.VARIANTS[meta].getColor(world, pos);
            }
        }
        return 0xFFFFFFFF;
    };

    private final TextureAtlasSprite[] destroyStageSprites = new TextureAtlasSprite[10];

    @Override
    public void onResourceManagerReload(IResourceManager rm) {
        Glazed.LOGGER.debug("Reloading destroy stage sprite cache");
        Arrays.setAll(destroyStageSprites, i -> ModelLoader.defaultTextureGetter()
                .apply(new ResourceLocation("minecraft:blocks/destroy_stage_" + i))
        );
    }

    @SubscribeEvent
    protected void onModelRegistry(ModelRegistryEvent event) {
        register(GlazedRegistry.GLASS_BLOCK, STATE_MAPPER_BLOCK);
        register(GlazedRegistry.GLASS_PANE, STATE_MAPPER_PANE);
        register(GlazedRegistry.KILN_BRICKS_ITEM, "normal");
        register(GlazedRegistry.GLASS_KILN_ITEM, "inventory");
        register(GlazedRegistry.GLASS_BLOCK_ITEM, "normal", GlazedVariant.VARIANTS);
        register(GlazedRegistry.GLASS_PANE_ITEM, "inventory", GlazedVariant.VARIANTS);
    }

    @SubscribeEvent
    protected void onBlockColors(ColorHandlerEvent.Block event) {
        register(event, GlazedRegistry.GLASS_BLOCK, BLOCK_COLOR);
        register(event, GlazedRegistry.GLASS_PANE, BLOCK_COLOR);
    }

    @SubscribeEvent
    protected void onItemColors(ColorHandlerEvent.Item event) {
        register(event, GlazedRegistry.GLASS_BLOCK_ITEM, ITEM_COLOR);
        register(event, GlazedRegistry.GLASS_PANE_ITEM, ITEM_COLOR);
    }

    @SubscribeEvent
    protected void onRenderWorldLast(RenderWorldLastEvent event) {
        final Minecraft mc = FMLClientHandler.instance().getClient();
        final PlayerControllerMP controller = mc.playerController;

        if (controller == null || !controller.getIsHittingBlock()) return;

        final BlockPos curr = controller.currentBlock;
        final IBlockState state = mc.world.getBlockState(curr);

        if (state.getBlock() != GlazedRegistry.GLASS_KILN) return;

        final int destroyStage = (int) (controller.curBlockDamageMP * 10.0F) - 1;

        if (destroyStage < 0 || destroyStage > 10) return;

        try {
            GlStateManager.tryBlendFuncSeparate(DST_COLOR, SRC_COLOR, ONE, ZERO);
            GlStateManager.enableBlend();
            GlStateManager.color(1.0F, 1.0F, 1.0F, 0.5F);
            GlStateManager.enablePolygonOffset();
            GlStateManager.doPolygonOffset(-3.0F, -3.0F);
            GlStateManager.enableAlpha();
            GlStateManager.alphaFunc(GL11.GL_GREATER, 0.1F);

            final Tessellator tessellator = Tessellator.getInstance();
            final BufferBuilder buffer = tessellator.getBuffer();
            final Vec3d vec = getPositionVector(mc.player, event.getPartialTicks());
            final TextureAtlasSprite sprite = destroyStageSprites[destroyStage];
            final BlockPos pos = state.getValue(GlassKilnBlock.HALF).offsetToOtherHalf(curr);

            buffer.begin(GL11.GL_QUADS, DefaultVertexFormats.BLOCK);
            buffer.setTranslation(-vec.x, -vec.y, -vec.z);
            buffer.noColor();

            mc.renderEngine.bindTexture(TextureMap.LOCATION_BLOCKS_TEXTURE);
            mc.getBlockRendererDispatcher().renderBlockDamage(state, pos, sprite, mc.world);

            buffer.setTranslation(0.0D, 0.0D, 0.0D);
            tessellator.draw();
        } catch (Exception e) {
            throw new RuntimeException("Encountered exception whilst rendering destroy stage quads", e);
        } finally {
            GlStateManager.tryBlendFuncSeparate(SRC_ALPHA, ONE_MINUS_SRC_ALPHA, ONE, ZERO);
            GlStateManager.doPolygonOffset(0.0F, 0.0F);
            GlStateManager.disablePolygonOffset();
            GlStateManager.depthMask(true);
        }
    }

    private Vec3d getPositionVector(Entity entity, float partialTicks) {
        final double lastX = entity.lastTickPosX;
        final double lastY = entity.lastTickPosY;
        final double lastZ = entity.lastTickPosZ;
        final double posX = lastX + ((entity.posX - lastX) * partialTicks);
        final double posY = lastY + ((entity.posY - lastY) * partialTicks);
        final double posZ = lastZ + ((entity.posZ - lastZ) * partialTicks);
        return new Vec3d(posX, posY, posZ);
    }

    private void register(Block block, IStateMapper stateMapper) {
        final ResourceLocation id = requireNonNull(block.getRegistryName());
        Glazed.LOGGER.debug("Registering state mapper for {}", id);
        ModelLoader.setCustomStateMapper(block, stateMapper);
    }

    private void register(Item item, String variant) {
        final ResourceLocation id = requireNonNull(item.getRegistryName());
        Glazed.LOGGER.debug("Registering item model for {}#0 to {}#{}", id, id, variant);
        ModelLoader.setCustomModelResourceLocation(item, 0, new ModelResourceLocation(id, variant));
    }

    private void register(Item item, String variant, IStringSerializable... variants) {
        final ResourceLocation id = requireNonNull(item.getRegistryName());
        for (int meta = 0; meta < variants.length; meta++) {
            final String path = variants[meta].getName() + "_" + id.getResourcePath();
            final ResourceLocation model = new ResourceLocation(Glazed.ID, path);
            Glazed.LOGGER.debug("Registering item model for {}#{} to {}#{}", id, meta, model, variant);
            ModelLoader.setCustomModelResourceLocation(item, meta, new ModelResourceLocation(model, variant));
        }
    }

    private void register(ColorHandlerEvent.Item event, Item item, IItemColor color) {
        final String cls = requireNonNull(color).getClass().getSimpleName();
        final ResourceLocation id = requireNonNull(item.getRegistryName());
        Glazed.LOGGER.debug("Registering item color instance {} for {}", cls, id);
        event.getItemColors().registerItemColorHandler(color, item);
    }

    private void register(ColorHandlerEvent.Block event, Block block, IBlockColor color) {
        final String cls = requireNonNull(color).getClass().getSimpleName();
        final ResourceLocation id = requireNonNull(block.getRegistryName());
        Glazed.LOGGER.debug("Registering block color instance {} for {}", cls, id);
        event.getBlockColors().registerBlockColorHandler(color, block);
    }
}

File: src/main/kotlin/net/insomniakitten/glazed/GlazedClient.kt
package net.insomniakitten.glazed

import net.insomniakitten.glazed.GlazedVariant.VARIANTS
import net.insomniakitten.glazed.block.GlassKilnBlock
import net.insomniakitten.glazed.extensions.get
import net.insomniakitten.glazed.extensions.getPosition
import net.insomniakitten.glazed.extensions.key
import net.insomniakitten.glazed.extensions.variant
import net.minecraft.block.Block
import net.minecraft.block.BlockPane.EAST
import net.minecraft.block.BlockPane.NORTH
import net.minecraft.block.BlockPane.SOUTH
import net.minecraft.block.BlockPane.WEST
import net.minecraft.block.state.IBlockState
import net.minecraft.client.renderer.GlStateManager.DestFactor
import net.minecraft.client.renderer.GlStateManager.SourceFactor
import net.minecraft.client.renderer.GlStateManager.alphaFunc
import net.minecraft.client.renderer.GlStateManager.color
import net.minecraft.client.renderer.GlStateManager.depthMask
import net.minecraft.client.renderer.GlStateManager.disablePolygonOffset
import net.minecraft.client.renderer.GlStateManager.doPolygonOffset
import net.minecraft.client.renderer.GlStateManager.enableAlpha
import net.minecraft.client.renderer.GlStateManager.enableBlend
import net.minecraft.client.renderer.GlStateManager.enablePolygonOffset
import net.minecraft.client.renderer.GlStateManager.tryBlendFuncSeparate
import net.minecraft.client.renderer.Tessellator
import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.client.renderer.block.statemap.IStateMapper
import net.minecraft.client.renderer.block.statemap.StateMapperBase
import net.minecraft.client.renderer.color.IBlockColor
import net.minecraft.client.renderer.color.IItemColor
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.client.renderer.texture.TextureMap
import net.minecraft.client.renderer.vertex.DefaultVertexFormats
import net.minecraft.client.resources.IResourceManager
import net.minecraft.client.resources.IResourceManagerReloadListener
import net.minecraft.item.Item
import net.minecraft.util.IStringSerializable
import net.minecraft.util.ResourceLocation
import net.minecraftforge.client.event.ColorHandlerEvent
import net.minecraftforge.client.event.ModelRegistryEvent
import net.minecraftforge.client.event.RenderWorldLastEvent
import net.minecraftforge.client.model.ModelLoader
import net.minecraftforge.fml.client.FMLClientHandler
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import org.lwjgl.opengl.GL11

@SideOnly(Side.CLIENT)
object GlazedClient : IResourceManagerReloadListener {
    private val BLOCK_STATE_MAPPER = object : StateMapperBase() {
        override fun getModelResourceLocation(state: IBlockState) = state.run {
            ModelResourceLocation("${Glazed.ID}:${key}_glass_block", "normal")
        }
    }

    private val PANE_STATE_MAPPER = object : StateMapperBase() {
        override fun getModelResourceLocation(state: IBlockState) = state.let {
            ModelResourceLocation(
                    "${Glazed.ID}:${it.key}_glass_pane",
                    "east=${it[EAST]},north=${it[NORTH]}," +
                    "south=${it[SOUTH]},west=${it[WEST]}"
            )
        }
    }

    private val BLOCK_COLOR = IBlockColor { state, access, pos, tint ->
        if (tint == 0 && access != null && pos != null) {
            state.variant.color(access, pos)
        }
        -1
    }

    private val ITEM_COLOR = IItemColor { stack, tint ->
        if (tint == 0 && stack in GlazedVariant) {
            client.player?.run { stack.variant?.run { color(world, position) } }
        }
        -1
    }

    private val client get() = FMLClientHandler.instance().client
    private val textureMap get() = client.textureMapBlocks
    private val textureManager get() = client.renderEngine
    private val blockRenderer get() = client.blockRendererDispatcher

    private var destroyStageSprites = emptyArray<TextureAtlasSprite>()

    override fun onResourceManagerReload(rm: IResourceManager) {
        Glazed.LOGGER.debug("Reloading destroy stage sprite cache")
        destroyStageSprites = Array(10) {
            textureMap.getAtlasSprite("minecraft:blocks/destroy_stage_$it")
        }
    }

    @SubscribeEvent
    fun onModelRegistry(event: ModelRegistryEvent) {
        register(GlazedRegistry.GLASS_BLOCK, BLOCK_STATE_MAPPER)
        register(GlazedRegistry.GLASS_PANE, PANE_STATE_MAPPER)
        register(GlazedRegistry.KILN_BRICKS_ITEM, "normal")
        register(GlazedRegistry.GLASS_KILN_ITEM, "inventory")
        register(GlazedRegistry.GLASS_BLOCK_ITEM, "normal", VARIANTS)
        register(GlazedRegistry.GLASS_PANE_ITEM, "inventory", VARIANTS)
    }

    @SubscribeEvent
    fun onBlockColors(event: ColorHandlerEvent.Block) {
        register(event, GlazedRegistry.GLASS_BLOCK, BLOCK_COLOR)
        register(event, GlazedRegistry.GLASS_PANE, BLOCK_COLOR)
    }

    @SubscribeEvent
    fun onItemColors(event: ColorHandlerEvent.Item) {
        register(event, GlazedRegistry.GLASS_BLOCK_ITEM, ITEM_COLOR)
        register(event, GlazedRegistry.GLASS_PANE_ITEM, ITEM_COLOR)
    }

    @SubscribeEvent
    fun onRenderWorldLast(event: RenderWorldLastEvent) {
        val controller = client.playerController ?: return

        if (!controller.isHittingBlock) return

        val curr = controller.currentBlock
        val state = client.world[curr].state

        if (state.block != GlazedRegistry.GLASS_KILN) return

        val destroyStage = (controller.curBlockDamageMP * 10.0f).toInt() - 1

        if (destroyStage < 0 || destroyStage > 10) return

        try {
            tryBlendFuncSeparate(
                    SourceFactor.DST_COLOR,
                    DestFactor.SRC_COLOR,
                    SourceFactor.ONE,
                    DestFactor.ZERO
            )
            enableBlend()
            color(1.0f, 1.0f, 1.0f, 0.5f)
            enablePolygonOffset()
            doPolygonOffset(-3.0f, -3.0f)
            enableAlpha()
            alphaFunc(GL11.GL_GREATER, 0.1f)

            val buffer = Tessellator.getInstance().buffer
            val posLerp = client.player.getPosition(event.partialTicks)
            val sprite = destroyStageSprites[destroyStage]
            val pos = state[GlassKilnBlock.HALF](curr)

            buffer.begin(GL11.GL_QUADS, DefaultVertexFormats.BLOCK)
            buffer.setTranslation(-posLerp.x, -posLerp.y, -posLerp.z)
            buffer.noColor()

            textureManager.bindTexture(TextureMap.LOCATION_BLOCKS_TEXTURE)
            blockRenderer.renderBlockDamage(state, pos, sprite, client.world)

            buffer.setTranslation(0.0, 0.0, 0.0)
            Tessellator.getInstance().draw()
        } catch (e: Exception) {
            throw RuntimeException("Encountered exception whilst rendering destroy stage quads", e)
        } finally {
            tryBlendFuncSeparate(
                    SourceFactor.SRC_ALPHA,
                    DestFactor.ONE_MINUS_SRC_ALPHA,
                    SourceFactor.ONE,
                    DestFactor.ZERO
            )
            doPolygonOffset(0.0f, 0.0f)
            disablePolygonOffset()
            depthMask(true)
        }
    }

    private fun register(block: Block, stateMapper: IStateMapper) {
        Glazed.LOGGER.debug("Registering state mapper for {}", block.registryName!!)
        ModelLoader.setCustomStateMapper(block, stateMapper)
    }

    private fun register(item: Item, variant: String) {
        val id = item.registryName!!
        Glazed.LOGGER.debug("Registering item model for {}#0 to {}#{}", id, id, variant)
        ModelLoader.setCustomModelResourceLocation(item, 0, ModelResourceLocation(id, variant))
    }

    private fun register(item: Item, variant: String, variants: Array<out IStringSerializable>) =
            variants.forEachIndexed { meta, it ->
                val id = item.registryName!!
                val model = ResourceLocation(Glazed.ID, it.name + "_" + id.resourcePath)
                Glazed.LOGGER.debug("Registering item model for {}#{} to {}#{}", id, meta, model, variant)
                ModelLoader.setCustomModelResourceLocation(item, meta, ModelResourceLocation(model, variant))
            }

    private fun register(event: ColorHandlerEvent.Item, item: Item, color: IItemColor) {
        Glazed.LOGGER.debug(
                "Registering item color instance {} for {}",
                color::class.java.simpleName, item.registryName!!
        )
        event.itemColors.registerItemColorHandler(color, item)
    }

    private fun register(event: ColorHandlerEvent.Block, block: Block, color: IBlockColor) {
        Glazed.LOGGER.debug(
                "Registering block color instance {} for {}",
                color::class.java.simpleName, block.registryName!!
        )
        event.blockColors.registerBlockColorHandler(color, block)
    }
}
